### PR TITLE
Add RebalancingInterceptor in misk to randomly close connections

### PR DESCRIPTION
**Pull request by jjestrel**
with a configurable percent chance

### CODE REVIEW

File: misk/src/main/kotlin/misk/web/WebConfig.kt [hunk 1]
```diff
@@ -14,7 +14,9 @@ data class WebConfig(
   val acceptors: Int? = null,
   val queue_size: Int? = null,
   val jetty_max_thread_pool_size: Int? = null,
-  val action_exception_log_level: ActionExceptionLogLevelConfig = ActionExceptionLogLevelConfig()
+  val action_exception_log_level: ActionExceptionLogLevelConfig = ActionExceptionLogLevelConfig(),
+  // TODO(jayestrella): Add a sane default value to this.
+  val close_connection_percent: Double = 0.0
```
Comment (mightyguava) [inline misk/src/main/kotlin/misk/web/WebConfig.kt]: sessions is using 0.005 in prod at full traffic

File: misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+class RebalancingInterceptor @Inject constructor(
+  private val random: ThreadLocalRandom,
+  private val probability: Double
+) : NetworkInterceptor {
+  override fun intercept(chain: NetworkChain): Response<*> {
+    val response = chain.proceed(chain.request)
+
+    if (random.current().nextDouble() < probability) {
+      return response.copy(headers = response.headers.newBuilder()
+          .set("Connection", "close")
```
Comment (mightyguava) [inline misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt]: Does `"Connection": "close"` translate to `GOAWAY` in Jetty? `"Connection": "close"` isn't part of the http2 protocol. The Go http2 server explicitly added it at some point. Jetty doesn't seem to have similar support: https://github.com/eclipse/jetty.project/issues/2788

File: misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Singleton
+  class Factory @Inject internal constructor(
+    private val random: ThreadLocalRandom,
+    private val webConfig: WebConfig
+  ) : NetworkInterceptor.Factory {
+    init {
+      val percent = webConfig.close_connection_percent
+      logger.info {
+        "sending connection: close to ${String.format("%.2f", percent)}% of all responses"
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt]: /nit can you add quotes around connection: close?

File: misk/src/test/kotlin/misk/web/AbstractRebalancingTest.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          close_connection_percent = percent
+      )))
+    }
+  }
+}
+
+@MiskTest(startService = true)
+class RebalancingEnabledTest : AbstractRebalancingTest(100.0) {
+  override fun checkResponse(response: Response) {
+    assertThat(response.header("Connection")).isEqualTo("close")
```
Comment (wesleyk) [inline misk/src/test/kotlin/misk/web/AbstractRebalancingTest.kt]: nice test!

File: misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt (hunk 2 above)
Comment (jmuia) [inline misk/src/main/kotlin/misk/web/interceptors/RebalancingInterceptor.kt]: this is cool!

File: misk/src/test/kotlin/misk/web/AbstractRebalancingTest.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import misk.testing.MiskTestModule
+import misk.web.jetty.JettyService
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import javax.inject.Inject
+
+abstract class AbstractRebalancingTest(
```
Comment (jmuia) [inline misk/src/test/kotlin/misk/web/AbstractRebalancingTest.kt]: neat

**Comment by tirsen** [comment]
This didn't work right?

**Comment by jjestrel** [comment]
This should only work for http1

**Pull request merged**